docs/design/configs.md

```diff
@@ -0,0 +1,130 @@
+# Generic Config Object
+
+## Abstract
+
+A generic ``Config`` object to store command-line flags and environment variables
+that is used for configuration of ``Kubernetes`` components living on top of ``etcd``
+is proposed in this proposal.
+
+The main focus of this proposal to solve dynamic configuration problem of components 
+and encapsulate configuration information to increase overall system quality and to 
+create a flexible configuration model for ``Kubernetes`` components.
+
+## Problem
+
+Currently command-line flags and environment variables are used for component configurations in ``Kubernetes``.
+But, this approach is not suitable for a distributed computing environment. Specifically the 
+problems with the current approach are:
+
+* Hard to synchronize across replicated components.
+* Hard to change dynamically as needed by continuous services.
+* Hard to change in different versions that may support different configuration options.
+
+## Solution
+
+A new ``Config`` resource is proposed to solve the aforementioned problems 
+that is stored in ``etcd`` and distributed to interested/necessary components 
+through the API server. ``Config`` resource will be mounted as a new volume 
+and will be managed by a side-car container that communicates with the API server.
+
+The ``Config`` resource architecture will be very similar to that of ``Secrets``.
+However, the ``Config`` resource will be dynamic, meaning that different components consuming
+the same ``Config`` will be notified about the changes to that particular resource.
+
+### Advantages
+
+* Reusable across different components.
+* Easy distribution through the API server.
+* By leveraging the power of ``/watch API`` along with the new resource, we can change component configurations dynamically.
+* Layer of abstraction that gets rid of the global state currently used for command-line flags and environment variables.
+
+### API Resource
+
+A new resource for ``Config`` will be added to the ``API``:
+
+```go
+	type Config struct {
+		TypeMeta   `json:",inline"`
+		ObjectMeta `json:"metadata,omitempty"`
+
+		// Flags contains command-line flags.
+		Flags pflag.FlagSet `json:"flags,omitempty"`
```

Comment:
Sorry, I wasn't proposing bringing in external structs.  I was saying - config is like an API, you have to maintain and migrate and default it, so we are using the tools we have for APIs there as well.  If the proposed "config" solution is too generic, we have to build that other stuff on top of the generic solution.  What happens when we want to rename "minion-timeout" to "node-timeout"?  How do we document "node-timeout", etc.

----- Original Message -----
> > +* Reusable across different components.
> > +* Easy distribution through the API server.
> > +* By leveraging the power of ``/watch API`` along with the new resource,
> > we can change component configurations dynamically.
> > +* Layer of abstraction that gets rid of the global state currently used
> > for command-line flags and environment variables.
> > +
> > +### API Resource
> > +
> > +A new resource for ``Config`` will be added to the ``API``:
> > +
> > +```go
> > +	type Config struct {
> > +		TypeMeta   `json:",inline"`
> > +		ObjectMeta `json:"metadata,omitempty"`
> > +
> > +		// Flags contains command-line flags.
> > +		Flags pflag.FlagSet `json:"flags,omitempty"`
> 
> I am a little anxious about importing external structs into our API.  For
> example, any non-exported fields will cause trouble.  we should just define
> our own key/value struct.  We should also define a standard form for things
> like lists (maybe "use JSON snippets").
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6477/files#r32873796